go/vt/schemadiff/schema_diff_test.go

```diff
@@ -0,0 +1,657 @@
+				require.NotEqualf(t, originalSingleString, toSingleString(pdiffs), "in iteration %d", iteration)
+			}
+			iteration++
+			return false
+		})
+		assert.False(t, earlyBreak)
+		assert.Equal(t, 24, len(allPerms))
+	}
+	{
+		allPerms := map[string]bool{}
+		allDiffs := schemaDiff.allDiffs()[:]
+		originalSingleString := toSingleString(allDiffs)
+		earlyBreak := permutateDiffs(allDiffs, func(pdiffs []EntityDiff) (earlyBreak bool) {
+			// Single visit
+			allPerms[toSingleString(pdiffs)] = true
+			// First permutation should be the same as original
+			require.Equal(t, originalSingleString, toSingleString(pdiffs))
+			// early break; this callback function should not be invoked again
+			return true
+		})
+		assert.True(t, earlyBreak)
+		assert.Equal(t, 1, len(allPerms))
+	}
+}
+
+func TestSchemaDiff(t *testing.T) {
+	var (
+		createQueries = []string{
+			"create table t1 (id int primary key, info int not null);",
+			"create table t2 (id int primary key, ts timestamp);",
+			"create view v1 as select id from t1",
+		}
+	)
+	tt := []struct {
+		name            string
+		fromQueries     []string
+		toQueries       []string
+		expectDiffs     int
+		expectDeps      int
+		sequential      bool
+		impossibleOrder bool
+		entityOrder     []string // names of tables/views in expected diff order
+	}{
+		{
+			name:        "no change",
+			toQueries:   createQueries,
+			entityOrder: []string{},
+		},
+		{
+			name: "three unrelated changes",
+			toQueries: []string{
+				"create table t1 (id int primary key, info int not null, ts timestamp);",
+				"create table t2 (id int primary key, ts timestamp, v varchar);",
+				"create view v1 as select id from t1",
+				"create view v2 as select 1 from dual",
+			},
+			expectDiffs: 3,
+			entityOrder: []string{"t1", "t2", "v2"},
+		},
+		{
+			name: "three unrelated changes 2",
+			toQueries: []string{
+				"create table t1 (id int primary key, info int not null);",
+				"create table t2 (id int primary key, ts timestamp, v varchar);",
+				"create view v2 as select 1 from dual",
+			},
+			expectDiffs: 3,
+			entityOrder: []string{"v1", "t2", "v2"},
+		},
+		// Subsequent
+		{
+			name: "add one fulltext key",
+			toQueries: []string{
+				"create table t1 (id int primary key, info int not null);",
+				"create table t2 (id int primary key, ts timestamp, v varchar, fulltext key ftk1 (v));",
+				"create view v1 as select id from t1",
+			},
+			expectDiffs: 1,
+			expectDeps:  0,
+			entityOrder: []string{"t2"},
+		},
+		{
+			name: "add two fulltext keys",
+			toQueries: []string{
+				"create table t1 (id int primary key, info int not null);",
+				"create table t2 (id int primary key, ts timestamp, v varchar, fulltext key ftk1 (v), fulltext key ftk2 (v));",
+				"create view v1 as select id from t1",
+			},
+			expectDiffs: 2,
+			expectDeps:  1,
+			sequential:  true,
+			entityOrder: []string{"t2", "t2"},
```

Comment:
I remember why from our previous discussions, but probably worth a comment why we expect this.

Reply:
Clarified in code comment.